pdr_backend/lake/partitioned_data_store.py

```diff
@@ -0,0 +1,114 @@
+    @enforce_types
+    def __init__(self, base_directory=str):
+        self.base_directory = base_directory
+        self.duckdb_conn = duckdb.connect(
+            database=":memory:"
+        )  # Keep a persistent connection
+
+    @enforce_types
+    def _generate_view_name(self, base_path=str) -> str:
+        """
+        Generate a unique view name for a given base path.
+        @arguments:
+            base_path - The base path to generate a view name for.
+        @returns:
+            str - A unique view name.
+        """
+
+        hash_object = hashlib.md5(base_path.encode())
+        return f"dataset_{hash_object.hexdigest()}"
+
+    @enforce_types
+    def append_data(self, df: pl.DataFrame, dataset_identifier: str):
+        """
+        Append data to an in-memory dataset.
+        @arguments:
+            df - The Polars DataFrame to append.
+            dataset_identifier - A unique identifier for the dataset.
+        """
+
+        view_name = self._generate_view_name(self.base_directory + dataset_identifier)
+        # Register the DataFrame as an Arrow table temporarily
+        # arrow_table = df.to_arrow()
+        self.duckdb_conn.register(view_name, df)
+
+    @enforce_types
+    def finalize_data(self, dataset_identifier: str):
+        """
+        Finalizes the data by writing the in-memory dataset to partitioned Parquet files.
+        @arguments:
+            dataset_identifier - A unique identifier for the dataset.
+        """
+
+        view_name = self._generate_view_name(self.base_directory + dataset_identifier)
+        query = f"SELECT * FROM {view_name}"
+        result_df = self.duckdb_conn.execute(query).df()
+
+        # Convert to Polars DataFrame for partitioned write
+        df = pl.DataFrame(result_df)
+
+        lazy_df = df.lazy()
+        # Determine partitions based on a timestamp column, assuming 'timestamp' column exists
+        lazy_df = lazy_df.with_columns(
+            [pl.from_epoch("timestamp", time_unit="s").alias("date")]
+        )
+        df = lazy_df.collect()
+
+        partitions = (
+            lazy_df.select(
+                pl.col("date").dt.year().alias("year"),
+                pl.col("date").dt.month().alias("month"),
+                pl.col("date").dt.day().alias("day"),
+            )
+            .unique()
+            .collect()
+        )
+
+        for partition in partitions.rows():
+            year, month, day = partition
+            partition_path = os.path.join(
+                self.base_directory,
+                dataset_identifier,
+                f"year={year}",
+                f"month={month}",
+            )
+            os.makedirs(partition_path, exist_ok=True)
+            partition_df = df.filter(
+                (pl.col("date").dt.year() == year)
+                & (pl.col("date").dt.month() == month)
+            )
+            file_path = os.path.join(partition_path, f"{day}.parquet")
+            partition_df.write_parquet(file_path)
```

Comment:
I think if all the data is already inside of duck_db through append(), then we can use duck db's partitioned-writes functionality to handle all of this, such that it doesn't have to be done twice: (1) read from duck_db (2) partition to disk

https://duckdb.org/docs/archive/0.9.2/data/partitioning/partitioned_writes

Reply:
Unfortunately we can not do that.

we can extract the day, month and year values as following, so it shortens our current method

`SELECT *, 
    EXTRACT(year FROM timestamp) AS year,
    EXTRACT(month FROM timestamp) AS month,
    EXTRACT(day FROM timestamp) AS day`
    
    
    However, the COPY command is not supporting this kind of manipulation on the data

Reply:
It is changed now